### PR TITLE
Generic StyleFactory to support span style for compose compatibility

### PR DESCRIPTION
**Pull request by folab33**

### CODE REVIEW

**Comment by markusressel** [comment]
Hey @folab33 ! :nerd_face: 
Looks like you put a lot of effort into this one :astonished: 
Before I take a deep dive into the code, could you give me a short description of what you did? :thinking:

**Comment by folab33** [comment]
> Hey @folab33 ! 🤓 Looks like you put a lot of effort into this one 😲 Before I take a deep dive into the code, could you give me a short description of what you did? 🤔

Jetpack compose doesn't work with SpannedString but with AnnotatedString. I made StyleFactory generic for it to work with other style like SpanStyle which will be use to produce annotated string in compose.

**Comment by markusressel** [comment]
Dies this still work with SpannedString though? Or is it compose only?

**Comment by folab33** [comment]
It works with both SpannedString and compose. I modified SpannableHighlighter to make it work with the existing implementation and I added a new class AnnotatedStringHighlighter to produces annotated string that can be used in compose.

File: app/src/main/AndroidManifest.xml [hunk 1]
```diff
@@ -9,7 +9,8 @@
         android:roundIcon="@mipmap/ic_launcher_round"
         android:supportsRtl="true"
         android:theme="@style/AppTheme">
-        <activity android:name=".MainActivity">
+        <activity android:name=".MainActivity"
+            android:exported="true">
```
Comment (markusressel) [inline app/src/main/AndroidManifest.xml:13]: Is this necessary?

File: build.gradle [hunk 2]
```diff
@@ -8,13 +8,13 @@ buildscript {
 
         gradle_plugin_version = '4.2.1'
 
-        minSdkVersion = 16
-        versionName = "3.1.0"
+        minSdkVersion = 21
```
Comment (markusressel) [inline build.gradle:11]: I am fine with upping the minSdk, but this has to become a major release bump then. just make 4.0.0 right away.

**Review by markusressel — CHANGES_REQUESTED** [review]
Looks good otherwise!

I am not too fond of the fact that Compose and legacy ColorSchemes have to be maintained separately now, but I also dont really see a better way...

File: app/src/main/AndroidManifest.xml (hunk 1 above)
Comment (folab33) [reply app/src/main/AndroidManifest.xml:13]: Yes, apps targeting android 12 and above must declare this in activity manifest

**Comment by markusressel** [comment]
I will have to make some tests on my device before merging, but looks good :+1: 
Thx!

**Pull request merged**